### DISCUSSION TITLE
review/consider about providing commonjs bundle whennpm install???

### DISCUSSION THREAD

**Discussion by lancety** [geckosio/geckos.io#200] [Ideas]
The readme mentioned now we are using es module and configed everything es module because many mdoules using es module. 

The issue is - there are many cases and popular modules still doing things the commonjs way. For making things easier, in lots of cases, developers dont need to bundle node_modules in the final output package, when the package working under commonjs mode, it uses node_modules, and the npm installed geckos package has 'type': 'module', so developers have to make their way to convert, and because this module used binary '.node' files, it made the process little more complex as well.

For example: I have a server app built by using vercel/pkg module, this module does not support packing es module, so I compiled my server code to commonjs module, but then I have to convert geckos.io common and server to commonjs module.  The issue is - for sharing code of client side and code of git submodules, the 'common' and 'client' geckos module need to be es module, so things become complex.

So what I think we can do is - client side is normally es module , which is ok.   Server side might be es module, or have to be commonjs module, how about when publish to npm we have commonjs support for server side code, or both es module and commonjs module version? 

Or maybe there are easy way for this use case I dont know.   
:D

**Comment by trekze** [comment]
hey @lancety . How did you convert geckos to cjs? I also use pkg to deploy server-code, where dynamic imports of ES modules is not supported. Thanks in advance.

**Reply by lancety** [reply]
I am using typescript to transpile code to cjs (tsconfig file listed below)
then use pkg to pack the transpiled files.  

tsconfig.json
```
{
    "target": "es2019",
    "lib": ["es2019"],
    "moduleResolution": "node",
    "module": "commonjs",

    "sourceMap": true,
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
  "compilerOptions": {
    "outDir": "../dist/server",
    "module": "commonjs",
    "skipLibCheck": true
  },
  // server entry, terrain worker & node server worker
  "include": [
    "index.ts",
    "index.dev.ts",
    "../gameBase/**/*.worker.node.ts",
    "../gameBase/gameConstFactory/**/*.worker.ts"
  ],
  "lib": [
    "es2019"
  ]
}
```


package.json of the server index.js root folder
```
{
  "name": "galaxy108-multiplayer-server",
  "version": "1.0.0",
  "description": "galaxy108-multiplayer-server",
  "private": true,
  "pkg": {
    "scripts": [
      "../dist/server/**/*.js"   // this is the transpiled files
    ],
    "assets": [
      "../gameServer/pkg/scriptLib/*.js",
      "../node_modules/node-datachannel/lib/*.js",
      "../xlsx/scriptBuffer/*.buf",
      "./greenlock.d/config.json",
      "./greenlockrc"
    ],
    "targets": [ "node14-win-x64", "node14-linux-x64" ],
    "outputPath": "gameServer/pkg"
  },
  "bin": "../dist/server/gameServer/index.js"     // this is the transpiled entry file (the index.ts in source folder)
}
```
